### ISSUE TITLE
Image corruption

### ISSUE THREAD

**Issue by mingard** [dadi/publish#361]
We've had reports of regular image corruption on upload. More investigation is necessary to determine whether Publish or API are the cause.

![screen_shot_2018-01-16_at_11 12 15_am](https://user-images.githubusercontent.com/1352469/35682087-7bf82264-072d-11e8-809c-49dd87370231.png)

**Comment by danwdart** [comment]
I can't save images at all so I think I need to backtrack:

```

------WebKitFormBoundaryHqTkYmLDfZryFgTe
Content-Disposition: form-data; name="file"

[object Object]
------WebKitFormBoundaryHqTkYmLDfZryFgTe--
```

**Comment by mingard** [comment]
@danwdart What sort are you testing with? To my knowledge, jpg is the only current supported format in Publish.

**Comment by danwdart** [comment]
I was testing PNG just there, the output of a JPEG was
```
------WebKitFormBoundarykSIAAD0PzmYsiBpv
Content-Disposition: form-data; name="file"; filename="JPEG_20180705_033115.jpg"
Content-Type: image/jpeg


------WebKitFormBoundarykSIAAD0PzmYsiBpv--
```

which looks slightly better.

However, we've still got a fail, so looking into why that could be.

**Comment by danwdart** [comment]
Okay - on dev this worked, on the test instance it didn't, so I'm suspecting some kind of cross domain upload blocking, perhaps, or just not being up to date.

There also seems to be a difference between saving an uploaded image and saving a cache of a localStorage uploaded image - the filename has gone.

So it seems I'm given a URL in the JSON output of the uploader (good) but Publish renders this URL with an [object Object] instead of some of the middle.

Makes me want to consider making new objects that throw errors if you try to `toString` them by default, that'd be much more handy

**Comment by danwdart** [comment]
OK, everything uploads properly in images properly with the following fix:
https://github.com/dadi/cloud/pull/178

**Comment by danwdart** [comment]
Now will see if I get the corruption by uploading a different image type.

**Comment by danwdart** [comment]
Weird to see an upload of one image accepts multiple uploaded files - is this intentional?

**Comment by danwdart** [comment]
PNGs seem fine too. Uploading a gif works fine, but then CDN's output is 

```
Content-Type: application/json

"Invalid output format"
```

Still can't get anything to glitch out like the above though. Sure it's not just a malformed source?

**Comment by mingard** [comment]
@danwdart are you uploading two images to the same field at the same time. This is the only way to consistently see it.

**Comment by mingard** [comment]
I found that if I used the same pictures as had previously updated okay, it didn't happen. I would go and find two new jpgs with unique names and try uploading at the same time. I also found that it happened more when API was local to publish, almost like some tmp file issue.

**Comment by danwdart** [comment]
So, using filestore CDN rather than using S3 CDN made it happen for you? I have that setup so I can try this afternoon.

**Comment by mingard** [comment]
I was using S3, but the machine running Publish was also running API.

**Comment by danwdart** [comment]
Ah right - well I normally run API, Publish and CDN on the same machine
Let's see if I can properly reproduce this.

**Comment by mingard** [comment]
I'd suggest deploying to an AWS instance too. I haven't tried to reproduce this on Unix or Windows.

**Comment by danwdart** [comment]
I've used the default S3, did you mean to start and use an EC2?

**Comment by mingard** [comment]
Currently I can replicated consistently when:
- Hosting on EC2
- Running API and Publish on the same instance
- Using S3 as a storage type

**Comment by danwdart** [comment]
Just sorting out a server with Francesco to test this with.

**Comment by danwdart** [comment]
I apparently will need VPN access

**Comment by danwdart** [comment]
Just sorting out this with filestore before I upgrade. Are there some S3 creds I can use?

**Comment by danwdart** [comment]
How far did you get - was this fixed @mingard ?

**Comment by jimlambie** [comment]
Fixed in API 3.2.2 & 4.03

* 3.2.2: `npm i @dadi/api@three`
* 4.0.3: `npm i @dadi/api@latest`
